okhttp/src/main/java/okhttp3/internal/http2/Hpack.kt

```diff
@@ -13,576 +13,584 @@
+          val doubled = arrayOfNulls<Header>(dynamicTable.size * 2)
+          System.arraycopy(dynamicTable, 0, doubled, dynamicTable.size, dynamicTable.size)
+          nextHeaderIndex = dynamicTable.size - 1
+          dynamicTable = doubled
         }
-        index = nextHeaderIndex--;
-        dynamicTable[index] = entry;
-        headerCount++;
+        index = nextHeaderIndex--
+        dynamicTable[index] = entry
+        headerCount++
       } else { // Replace value at same position.
-        index += dynamicTableIndex(index) + entriesEvicted;
-        dynamicTable[index] = entry;
+        index += dynamicTableIndex(index) + entriesEvicted
+        dynamicTable[index] = entry
       }
-      dynamicTableByteCount += delta;
+      dynamicTableByteCount += delta
     }
 
-    private int readByte() throws IOException {
-      return source.readByte() & 0xff;
+    @Throws(IOException::class)
+    private fun readByte(): Int {
+      return source.readByte() and 0xff
     }
 
-    int readInt(int firstByte, int prefixMask) throws IOException {
-      int prefix = firstByte & prefixMask;
+    @Throws(IOException::class)
+    fun readInt(firstByte: Int, prefixMask: Int): Int {
+      val prefix = firstByte and prefixMask
       if (prefix < prefixMask) {
-        return prefix; // This was a single byte value.
+        return prefix // This was a single byte value.
       }
 
       // This is a multibyte value. Read 7 bits at a time.
-      int result = prefixMask;
-      int shift = 0;
+      var result = prefixMask
+      var shift = 0
       while (true) {
-        int b = readByte();
-        if ((b & 0x80) != 0) { // Equivalent to (b >= 128) since b is in [0..255].
-          result += (b & 0x7f) << shift;
-          shift += 7;
+        val b = readByte()
+        if (b and 0x80 != 0) { // Equivalent to (b >= 128) since b is in [0..255].
+          result += b and 0x7f shl shift
+          shift += 7
         } else {
-          result += b << shift; // Last byte.
-          break;
+          result += b shl shift // Last byte.
+          break
         }
       }
-      return result;
+      return result
     }
 
-    /** Reads a potentially Huffman encoded byte string. */
-    ByteString readByteString() throws IOException {
-      int firstByte = readByte();
-      boolean huffmanDecode = (firstByte & 0x80) == 0x80; // 1NNNNNNN
-      int length = readInt(firstByte, PREFIX_7_BITS);
+    /** Reads a potentially Huffman encoded byte string.  */
+    @Throws(IOException::class)
+    fun readByteString(): ByteString {
+      val firstByte = readByte()
+      val huffmanDecode = firstByte and 0x80 == 0x80 // 1NNNNNNN
+      val length = readInt(firstByte, PREFIX_7_BITS)
 
       if (huffmanDecode) {
-        return ByteString.of(Huffman.get().decode(source.readByteArray(length)));
+        val bytes = source.readByteArray(length.toLong())
+        return ByteString.of(*Huffman.get().decode(bytes))
```

Comment:
Does the splat (*) to a vararg method imply an extra copy?

Reply:
It does and that’s sad. Here’s the decompiled bytecode:

```
      @NotNull
      public final ByteString readByteString() throws IOException {
         int firstByte = this.readByte();
         boolean huffmanDecode = (firstByte & 128) == 128;
         int length = this.readInt(firstByte, 127);
         if (huffmanDecode) {
            byte[] bytes = this.source.readByteArray((long)length);
            Companion var10000 = ByteString.Companion;
            byte[] var10001 = Huffman.get().decode(bytes);
            Intrinsics.checkExpressionValueIsNotNull(var10001, "Huffman.get().decode(bytes)");
            return var10000.of(Arrays.copyOf(var10001, var10001.length));
         } else {
            return this.source.readByteString((long)length);
         }
      }
```

Reply:
... I changed to use a better API, `ByteArray.toByteString()` which saves the copy. Good catch!

```
      @NotNull
      public final ByteString readByteString() throws IOException {
         int firstByte = this.readByte();
         boolean huffmanDecode = (firstByte & 128) == 128;
         int length = this.readInt(firstByte, 127);
         if (huffmanDecode) {
            byte[] bytes = this.source.readByteArray((long)length);
            Companion var10000 = ByteString.Companion;
            byte[] var10001 = Huffman.get().decode(bytes);
            Intrinsics.checkExpressionValueIsNotNull(var10001, "Huffman.get().decode(bytes)");
            return Companion.of$default(var10000, var10001, 0, 0, 3, (Object)null);
         } else {
            return this.source.readByteString((long)length);
         }
      }
```

Reply:
... also fixed 3 similar splats in Cache, CertificatePinner, and HeldCertificate.